jmh-core/src/main/java/org/openjdk/jmh/profile/XCTraceTableProfileHandler.java

```diff
@@ -0,0 +1,373 @@
+                pushCachedOrNew(attributes, id -> new Frame(id, parseName(attributes),
+                        parseAddress(attributes) - 1L));
+                break;
+        }
+    }
+
+    @Override
+    public void endElement(String uri, String localName, String qName) {
+        if (qName.equals(XCTraceTableHandler.NODE)) {
+            return;
+        }
+        switch (qName) {
+            case XCTraceTableHandler.SAMPLE:
+                callback.accept(currentSample);
+                currentSample = null;
+                break;
+            case XCTraceTableHandler.SAMPLE_TIME: {
+                LongHolder value = popAndUpdateLongHolder();
+                currentSample.setTime(value.value);
+                break;
+            }
+            case XCTraceTableHandler.CYCLE_WEIGHT:
+            case XCTraceTableHandler.WEIGHT:
+            case XCTraceTableHandler.PMC_EVENT:
+                // in practice, a sample's row will contain only one of these
+                LongHolder value = popAndUpdateLongHolder();
+                currentSample.setWeight(value.value);
+                break;
+            case XCTraceTableHandler.BACKTRACE:
+                Frame topFrame = this.<ValueHolder<Frame>>pop().value;
+                currentSample.setTopFrame(topFrame.address, topFrame.name, topFrame.binary);
+                break;
+            case XCTraceTableHandler.BINARY:
+                ValueHolder<String> bin = pop();
+                this.<Frame>peek().binary = bin.value;
+                break;
+            case XCTraceTableHandler.FRAME:
+                Frame frame = pop();
+                ValueHolder<Frame> backtrace = peek();
+                // we only need a top frame
+                if (backtrace.value == null) {
+                    backtrace.value = frame;
+                }
+                break;
+            case XCTraceTableHandler.TEXT_ADDRESSES: {
+                LongHolder addresses = pop();
+                if (isNeedToParseCharacters()) {
+                    // peek only the first address as we're not interested in the whole backtrace
+                    addresses.value = Arrays.stream(getCharacters().split(" "))
+                            .mapToLong(Long::parseUnsignedLong).findFirst().orElse(-1);
+                }
+                ValueHolder<Frame> bt = peek();
+                // For legacy backtraces, the address is initially -1. It is then updated by the top-most address
+                // extracted from text-addresses elements.
+                if (bt.value.address == -1 && addresses.value != -1) {
+                    bt.value.address = addresses.value;
+                }
+                break;
+            }
+
+        }
+        setNeedParseCharacters(false);
+    }
+
+    @Override
+    public void endDocument() {
+        entriesCache.clear();
+        entriesStack.clear();
+    }
+
+    private static abstract class TraceElement {
+        public final long id;
```

Comment:
Here and later in `TraceElement` subclasses. For extra sanity, can you turn these fields `private` and expose proper (immutable) getters for them?

Reply:
Sure! Also, encapsulated mutable fields.